npeps/npep-187-ports-and-protocols.md

```diff
@@ -0,0 +1,271 @@
+
+```yaml
+protocols:
+  tcp:
+    - port: 8080
+    - portRange: 
+        start: 8080
+        end: 9090
+  udp:
+    - port: 8080
+    - port: 9090
+  namedPort:
+    - http
+    - monitoring
+  icmp:
+    - type: 7
+      code: 3
+```
+
+with the validation like so:
+- empty `protocols` list is not allowed
+- at least 1 field of `protocol` must be set
+- each protocol list must not be empty
+- exactly one of `port` or `portRange` must be set for TCP/UDP/SCTP elements
+
+
+This option is easier to write, takes less space and simplifies the validation, but doesn't allow defaulting protocol to TCP 
+(which is the most popular case). On the other hand, I am not convinced that defaulting protocol to TCP is
+a) a good idea (less transparency on which protocol is used)
+b) much simpler than adding `tcp:`
+
+To compare, the first option for `tcp:8080` is
+```yaml
+protocols:
+  - port: 
+      number: 8080
+```
+and the second is
+```yaml
+protocols:
+  tcp:
+    - port: 8080
+```
+
+And if you want to match 3 TCP ports: 80, 8080, 443 (you can imagine how it looks with even more ports)
+
+```yaml
+protocols:
+  - port: 
+      number: 80
+  - port:
+      number: 8080
+  - port:
+      number: 443
+```
+vs
+```yaml
+protocols:
+  tcp:
+    - port: 80
+    - port: 8080
+    - port: 443
+```
+
+Allow DNS (TCD and UDP port 53):
+
+```yaml
+protocols:
+  - port: 
+      number: 53
+  - protocol: UDP
+    port:
+      number: 53
+```
+vs
+```yaml
+protocols:
+  tcp:
+    - port: 53
+  udp:
+    - port: 53
+```
+
+Allow a range of UDP ports (e.g. for telephony use cases):
+
+```yaml
+protocols:
+  - protocol: UDP
+    port:
+      range:
+        start: 10000
+        end: 20000
+```
+vs
+```yaml
+protocols:
+  udp:
+    - portRange:
+        start: 10000
+        end: 20000
+```
+
+### New protocols
+
+Using ICMP as a potential example for a pure protocol match (`protocol: ICMP`) or protocol with extra fields (e.g. ICMP type and code),
+a future extension could look like this:
+
+Option 1:
+```yaml
+protocols:
+  # option 1:  A little inconsistent with port match as it has an extra `icmp` field instead of flat parameters
+  - protocol: ICMP
+    icmp:
+      type: 7
+      code: 3
+  # option 2: Flat parameters, may conflict with the future protocol that also have type or code
+  - protocol: ICMP
+    type: 7
+    code: 3
+  # option 3: Same as option 1, but without a discriminator (protocol)
+  - icmp:
+      type: 7
+      code: 3
+```
+
+Option 2:
+```yaml
+protocols:
+  icmp:
+    - type: 7
+      code: 3
+```
+
+A likely option is that we will start with the `ICMP: match all` option, as more details matching is not very popular,
+but we also want to leave space for future extensions (adding more fields), so the struct should allow both
+
+Option 1:
+```yaml
+protocols:
+  - protocol: ICMP
+    icmp:
+      matchAll: true
```

Comment:
Is `icmp: {}` an option (was discussed on the Slack thread).

Reply:
Using an empty struct to represent something means that you can never ever add another way of selecting.  This is the same problem with old clients. If a new user specifies one of the newer selection modes, and an older implementation reads it and drops that field, it will assume that means "all", which, for Network policy, could mean failing open. 

It's better for "all" to be an explicit thing. 

In an ideal world, you could perhaps make "all" be the default, applied by the server. If nothing else is specified. Then clients are not subject to the empty structure.  That requires defaulting logic to look at multiple Fields, which isn't very well supported right now, unless you hand code the defaulting.

Reply:
Yes, I agree! Having an explicit enum type field description is IMO the most explicit and least error prone path.

In general, I think we should bias towards being more consistent and easy to understand with less emphasis on conciseness because most users will be generating their configuration using an (AI) tool.